source/common/quic/envoy_quic_client_stream.cc

```diff
@@ -80,30 +80,38 @@ void EnvoyQuicClientStream::encodeData(Buffer::Instance& data, bool end_stream)
   ASSERT(!local_end_stream_);
   local_end_stream_ = end_stream;
   SendBufferMonitor::ScopedWatermarkBufferUpdater updater(this, this);
-  Buffer::RawSliceVector raw_slices = data.getRawSlices();
-  absl::InlinedVector<quiche::QuicheMemSlice, 4> quic_slices;
-  quic_slices.reserve(raw_slices.size());
-  for (auto& slice : raw_slices) {
-    ASSERT(slice.len_ != 0);
-    // Move each slice into a stand-alone buffer.
-    // TODO(danzh): investigate the cost of allocating one buffer per slice.
-    // If it turns out to be expensive, add a new function to free data in the middle in buffer
-    // interface and re-design QuicheMemSliceImpl.
-    quic_slices.emplace_back(quiche::QuicheMemSlice::InPlace(), data, slice.len_);
-  }
-  quic::QuicConsumedData result{0, false};
-  absl::Span<quiche::QuicheMemSlice> span(quic_slices);
-  {
+  if (http_datagram_handler_) {
     IncrementalBytesSentTracker tracker(*this, *mutableBytesMeter(), false);
-    result = WriteBodySlices(span, end_stream);
-  }
-  // QUIC stream must take all.
-  if (result.bytes_consumed == 0 && has_data) {
-    IS_ENVOY_BUG(fmt::format("Send buffer didn't take all the data. Stream is write {} with {} "
-                             "bytes in send buffer. Current write was rejected.",
-                             write_side_closed() ? "closed" : "open", BufferedDataBytes()));
-    Reset(quic::QUIC_BAD_APPLICATION_PAYLOAD);
-    return;
+    if (!http_datagram_handler_->encodeCapsuleFragment(data.toString(), end_stream)) {
```

Comment:
Hmmm... So if an EnvoyQuicClientStream has a datagram handler, then any data to be encoded will be sent as datagrams? That feels a bit surprising to me, but maybe it shouldn't? At a minimum, we should add comments to http_datagram_handler_ in the .h files which explains what it means for this to be set.

Reply:
Once the client stream receives a 2xx status code response with a capsule-protocol header (or Upgrade: connect-udp), the datagram handler will be set by calling enableHttpDatagramSupport (in the next PR) and subsequent DATA frames will be encoded as Datagrams until the stream is reset.
I think this should be enough to correctly implement CONNECT-UDP. Let me know if you have any further questions or thoughts.